### PR TITLE
[experimental] added exact_h2 to ldscsim

### PR DESCRIPTION
**Pull request by nikbaya**
exact_h2 flag enables more precise h2's in simulated traits. Flag is not supported for multitrait simulation.

Also added _get_tid() for getting temporary ID strings to append to mt or ht field names inside of functions.

### CODE REVIEW

**Comment by tpoterba** [comment]
Will review this by the end of the week, Nik.

**Comment by tpoterba** [comment]
Sorry I didn't get to this, Nik! I've been trying to think if we should revisit our experimental module guidelines in light of the questions we're getting on Zulip, but I think things are probably fine.

File: hail/python/hail/experimental/ldscsim.py [hunk 1]
```diff
@@ -82,43 +85,51 @@ def simulate_phenotypes(mt, genotype, h2, pi=None, rg=None, annot=None, popstrat
...
     :class:`.MatrixTable`
         :class:`.MatrixTable` with simulated betas and phenotypes, simulated according
         to specified model.
     """
-    h2 = h2.tolist() if type(h2) is np.ndarray else ([h2] if type(h2) is not list else h2) 
+    h2 = h2.tolist() if type(h2) is np.ndarray else ([h2] if type(h2) is not list else h2)
     pi = pi.tolist() if type(pi) is np.ndarray else pi
-    tid = ''.join(random.choices(string.ascii_uppercase + string.ascii_lowercase, k=5)) # "temporary id" -- random string to identify temporary intermediate fields generated by this method
+    # tid = "temporary id" -- random string to identify temporary intermediate fields generated by this method
+    tid = _get_tid()
```
Comment (tpoterba) [inline hail/python/hail/experimental/ldscsim.py]: you can use `hl.utils.java.Env.get_uid()`, which we do internally in the library.

File: hail/python/hail/experimental/ldscsim.py [hunk 2]
```diff
@@ -508,69 +558,116 @@ def calculate_phenotypes(mt, genotype, beta, h2, popstrat=None, popstrat_var=Non
...
+    mt = mt.filter_rows(hl.agg.stats(mt['gt_'+tid]).stdev>0)
     mt = normalize_genotypes(mt['gt_'+tid])
-    if mt['beta_'+tid].dtype == dtype('array<float64>'): #if >1 traits
-        mt = mt.annotate_cols(y_no_noise = hl.agg.array_agg(lambda beta: hl.agg.sum(beta*mt['norm_gt']),mt['beta_'+tid]))
-        mt = mt.annotate_cols(y = mt.y_no_noise + hl.literal(h2).map(lambda x: hl.rand_norm(0,hl.sqrt(1-x))))
+    if mt['beta_'+tid].dtype == dtype('array<float64>'):  # if >1 traits
+        if exact_h2:
+            raise ValueError('exact_h2=True not supported for multitrait simulations')
+            # still working on this
+#            mt = mt.annotate_cols({'y_no_noise_'+tid: hl.agg.array_agg( #use tid because this is the unnormalized-for-h2 version of y_no_noise
```
Comment (tpoterba) [inline hail/python/hail/experimental/ldscsim.py]: prefer not to have large blocks of commented code -- delete and add implementation later?

File: hail/python/hail/experimental/ldscsim.py (hunk 1 above)
Comment (nikbaya) [reply hail/python/hail/experimental/ldscsim.py]: oh cool, I should have asked if you already had something like this

File: hail/python/hail/experimental/ldscsim.py (hunk 2 above)
Comment (nikbaya) [reply hail/python/hail/experimental/ldscsim.py]: Yeah no problem, we can delete this for now

**Comment by tpoterba** [comment]
got a doctest error:
```
______________ [doctest] hail.experimental.ldscsim.get_cov_matrix ______________
035     trait 3 & trait 4 :math:`r_g` = 0.6
036     To obtain the covariance matrix corresponding to this scenario :math:`h^2` values are
037     ordered according to user specification and :math:`r_g` values are ordered by the 
038     order in which the corresponding genetic covariance terms will appear in the 
039     covariance matrix, reading lines in the upper triangular matrix from left to
040     right, top to bottom (read first row left to right, read second row left to 
041     right, etc.), exluding the diagonal.
042 
043     >>> cov_matrix, rg = get_cov_matrix(h2=[0.1, 0.3, 0.2, 0.6], rg=[0.4, 0.3, 0.1, 0.2, 0.15, 0.6])
044     >>> print(cov_matrix)
Differences (unified diff with -expected +actual):
    @@ -1,4 +1,4 @@
    -array([[0.1       , 0.06928203, 0.04242641, 0.0244949 ],
    -    [0.06928203, 0.3       , 0.04898979, 0.06363961],
    -    [0.04242641, 0.04898979, 0.2       , 0.2078461 ],
    -    [0.0244949 , 0.06363961, 0.2078461 , 0.6       ]]
    +[[0.1        0.06928203 0.04242641 0.0244949 ]
    + [0.06928203 0.3        0.04898979 0.06363961]
    + [0.04242641 0.04898979 0.2        0.2078461 ]
    + [0.0244949  0.06363961 0.2078461  0.6       ]]

/hail/python/hail/experimental/ldscsim.py:44: DocTestFailure

```

File: hail/python/hail/experimental/ldscsim.py [hunk 3]
```diff
@@ -285,49 +317,62 @@ def multitrait_ss(mt, h2, pi, rg=0, seed=None):
...
         Seed for random number generator. If `seed` is ``None``, `seed` is set randomly.
-    
+
     Warning
     -------
     May give inaccurate results if chosen parameters make the covariance matrix 
     not positive semi-definite. Covariance matrix is likely to not be positive
-    semi-definite when :math:`p_{TT}` is small and rg is large.
-    
+    semi-definite when :mat:`p_{TT}` is small and rg is large.
```
Comment (tpoterba) [inline hail/python/hail/experimental/ldscsim.py]: typo: this is causing a docs build error

Comment (nikbaya) [reply hail/python/hail/experimental/ldscsim.py]: argh, sorry about that. It should now be fixed

**Comment by tpoterba** [comment]
I think you'll still have the doctest failure. It just returns list of lists now ,right?

**Comment by nikbaya** [comment]
> I think you'll still have the doctest failure. It just returns list of lists now ,right?

I'm not sure I follow. What returns a list of lists?

**Comment by tpoterba** [comment]
[this comment](https://github.com/hail-is/hail/pull/8265#issuecomment-601360575) describes the test failure -- it looks like something was a numpy array, but is now a nested list.

**Comment by nikbaya** [comment]
I think it was because when I print a numpy array it shows up as a list of lists.

**Comment by tpoterba** [comment]
Ugh, sorry, this has been failing unreliable CI/services tests. I'll keep bumping until it goes in

**Pull request merged**